core-util/v2/FunctionPointer.hpp

```diff
@@ -0,0 +1,110 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2015 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef __FUNCTIONPOINTER_HPP__
+#define __FUNCTIONPOINTER_HPP__
+
+#include <string.h>
+#include <stdint.h>
+#include <stddef.h>
+#include <stdarg.h>
+#include <new>
+#include <utility>
+#include <cstdio>
+#include <type_traits>
+
+namespace mbed {
+namespace util {
+
+
+namespace impl{
+
+template<size_t I = 0>
```

Comment:
```I``` is not optional padding. It is extra space for storing functors. Since bound arguments are captured in functors, this is equivalent to:

* Lambda capture list storage space
* Generic functor storage space
* Storage space for bound arguments

I don't think that ```PADDING``` captures these use-cases well.  Neither does ```I```.  I would suggest ```ExtraSpace```, but I prefer my template parameters to be more concise.  Suggestions welcome.